### ISSUE TITLE
LG-324: Add more fields in JSON message when an Exception is thrown

### ISSUE THREAD

**Issue by NickWemekamp** [OpenLiberty/open-liberty#17559]
**Is your feature request related to a problem? Please describe.**
We are using Open Liberty in a Kubernetes environment. We use JSON console logging to send logging to the ELK stack.
I want to log the Exception in a seperate field in the JSON log record, the message belonging to that error and the stacktrace.
So that I can search for Exceptions of a specific type. This is currently not possible because I can't parse the "message" field in Open Liberty

**Describe the solution you'd like**
`logger.info("test', new RuntimeException("error"))`

- Add field error.type which is RuntimeException
- Let message field only log the message of the error. So error.message would be 'test' in this example. Instead of 'test\njava.lang.RuntimeException (and the rest of the stack trace)'
- Add field error.stack_trace which has the entire stacktrace.

Having an error.type field would be very useful. The other 2 fields are nice to have but also useful. Names of the fields can of course be different.

**Describe alternatives you've considered**
Allow us to choose a Java Util Logging Formatter for the Open Liberty server. e.g. (https://github.com/elastic/ecs-logging-java)

**Additional context**
Most people that use JSON logging use it for a centralized log platform e.g. (ELK or Splunk). Having seperate fields for exception related data (instead of 1 field 'message') makes it easier to search for errors and configure alerting rules.

**Comment by Channyboy** [comment]
@NickWemekamp 
Tentatively, the aim is to duplicate the formatting done with the `liberty_ffdc` type.
The current format for `liberty_ffdc` provides:
`ibm_exceptionName` this is equivalent to your proposed `error.type`
`message` will be just be the message (i.e. the `test` in `logger.logp("test', new RuntimeException("error"))`)
(new) `ibm_exceptionMessage` will be the message from the throwable/exception
`ibm_stackTrace` this will be the full stack trace i.e. the `error.stack_trace`

It's best to use the _same_ keys so as not to duplicate and complicate the namespace.
The use of the `ibm_` prefix is to distinguish fields originating from Liberty and not conflict/clash with any other system that may be forwarding logs to the same logging platform.

**Comment by NickWemekamp** [comment]
@Channyboy yes looks good. So if I understood you correctly the `ibm_exceptionName`, `ibm_exceptionMessage` and `ibm_stackTrace` fields will be added to the JSON message event (https://openliberty.io/docs/21.0.0.11/json-log-events-list.html#_message_events). And the `message` field will not have the stacktrace anymore as that is then already in `ibm_stackTrace`.

**Comment by Channyboy** [comment]
@NickWemekamp Yup and the `error` in `logger.logp("test', new RuntimeException("error"))` will be part of the `ibm_stackTrace`

**Comment by Channyboy** [comment]
UFO link : https://ibm.box.com/s/wspgtfgrdk4pohq3dhkb79ze0r9awymj

## Description

JUL methods that log with a `Throwable` parameter well have the exception data parsed and presented as fields during JSON logging. The fields are `ibm_exceptionName` and `ibm_stackTrace`. They represent the exception name and stacktrace respectively.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Documents

When available, add links to required feature documents. Use "N/A" to mark particular documents which are not required by the feature.

- Aha: Externally raised RFE ([Aha](https://cloud-platform.ideas.aha.io/))
  - Link the RFE with this issue
- UFO: https://ibm.box.com/s/wspgtfgrdk4pohq3dhkb79ze0r9awymj
    - If you lack permissions, contact [OpenLiberty/release-architect](https://github.com/orgs/OpenLiberty/teams/release-architect)
- FTS: Link to Feature Test Summary GH Issue- https://github.com/OpenLiberty/open-liberty/issues/21369
- Beta Blog: Link to Beta Blog Post GH Issue - https://openliberty.io/blog/2022/07/07/beta-release-post.html
- GA Blog: Link to GA Blog Post GH Issue

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Process Overview

- [Prioritization](#prioritization)
- [Design](#design)
- [Implementation](#implementation)
- [Legal and Translation](#legal-and-translation)
- [Beta](#beta)
- [GA](#ga)
  - [Focal Point Approvals](#focal-point-approvals-complete-by-feature-complete-date)
- [Other Deliverables](#other-deliverables)

### General Instructions

The process steps occur roughly in the order as presented. Process steps occasionally overlap.

Each process step has a number of tasks which must be completed or must be marked as not applicable ("N/A").

Unless otherwise indicated, the tasks are the responsibility of the Feature Owner or a Delegate of the Feature Owner.

If you need assistance, reach out to the [OpenLiberty/release-architect](https://github.com/orgs/OpenLiberty/teams/release-architect).

**Important: Labels are used to trigger particular steps and must be added as indicated.** 

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Prioritization** (Complete Before Development Starts)

The ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) and area leads are responsible for prioritizing the features and determining which features are being actively worked on.

### **Prioritization**
- [x] Feature added to the "New" column of the [Open Liberty project board](https://github.com/orgs/OpenLiberty/projects/2)
  - Epics can be added to the board in one of two ways:
    - From this issue, use the "Projects" section to select the appropriate project board.
    - From the appropriate project board click "Add card" and select your Feature Epic issue
- [x] Priority assigned
  - Attend the Liberty Backlog Prioritization meeting

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Design** (Complete Before Development Starts)

Design preliminaries determine whether a formal design, which will be provided by an Upcoming Feature Overview (UFO) document, must be created and reviewed.  A formal design is required if the feature requires any of the following: UI, Serviceability, SVT, Performance testing, or non-trivial documentation/ID.

### **Design Preliminaries**
- [x] UI requirements identified. (Owner and [UI focal point](https://github.com/orgs/OpenLiberty/teams/ui-approvers))
- [x] ID requirements identified. (Owner and [ID focal point](https://github.com/orgs/OpenLiberty/teams/id-approvers))
   - Refer to [Documenting Open Liberty](https://github.com/OpenLiberty/open-liberty/wiki/Documenting-Open-Liberty).
   - Feature Owner adds label `ID Required`, if non-trivial documentation needs to be created by the ID team.
   - ID adds label `ID Required - Trivial`, if no design will be performed and only trivial ID updates are needed.
- [x] Serviceability Requirements Identified. (Owner and [Serviceability focal point](https://github.com/orgs/OpenLiberty/teams/serviceability-approvers))
- [x] SVT Requirements Identified. (Owner and [SVT focal point](https://github.com/orgs/OpenLiberty/teams/svt-approvers))
- [x] Performance testing requirements identified. (Owner and [Performance focal point](https://github.com/orgs/OpenLiberty/teams/performance-approvers))

### **Design**
- [x] POC Design / UFO review requested.
  - Owner adds label `Design Review Request`
- [x] POC Design / UFO review scheduled. (David Chang)
- [x] POC Design / UFO review completed.
- [x] POC / UFO Review follow-ons completed.
- [x] Design / UFO approved. ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) or N/A
  - ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) adds label `Design Approved`
  - Add the public link to the UFO in Box to the [Documents](#documents) section.
  - The UFO must always accurately reflect the final implementation of the feature. Any changes must be first approved. Afterwards, update the UFO by creating a copy of the original approved slide(s) at the end of the deck and prepend "OLD" to the title(s). A single updated copy of the slide(s) should take the original's place, and have its title(s) prepended with "UPDATED".

### **No Design**
- [ ] No Design requested.
  - Owner adds label `No Design Approval Request`
- [ ] No Design / No UFO approved. ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) or N/A
  - Approver adds label `No Design Approved`

### **FAT Documentation**
- [ ] "Feature Test Summary" child task created
  - Use the [Feature Test Summary Template](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=&labels=Feature+Test+Summary&template=feature_test_summary.md&title=)
  - Add FTS issue link to the [Documents](#documents) section.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Implementation**

A feature must be [prioritized](https://github.com/orgs/OpenLiberty/projects/2) and socialized (or `No Design Approved`) before any implementation work may begin and is the minimum before any beta code may be delivered.  All new Liberty content must be inaccessible in our GA releases until it is [Feature Complete](#feature-complete) by either marking it `kind=noship` or [beta fencing](#beta-code) it.  
Code may not GA until this feature has obtained the "Design Approved" or "No Design Approved" label, along with all other tasks outlines in [GA](#ga) section.

### **Feature Development Begins**
- [ ] Add the `In Progress` label

## **Legal and Translation**

In order to avoid last minute blockers and significant disruptions to the feature, the **legal items need to be done as early in the feature process as possible**, either in design or as early into the development as possible.  Similarly, translation is to be done concurrently with development.  Both **MUST** be completed before Beta or GA is requested.

### **Legal** (Complete before Feature Complete Date)
- [ ] Changed or new open source libraries are cleared and approved, or N/A. (Legal Release Services/Cass Tucker/Release PM).
- [ ] Licenses and Certificates of Originality (COOs) are updated, or N/A

### **Translation** (Complete 1 week before Feature Complete Date)
- [ ] PII updates are merged, or N/A. Note timing with translation shipments.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Beta**

In order to facilitate early feedback from users, all new features and functionality should first be released as part of a beta release.

### **Beta Code**
- [x] Beta fence the functionality
  - `kind=beta`, `ibm:beta`, `ProductInfo.getBetaEdition()`
- [x] Beta development complete and feature ready for inclusion in a beta release
  - Add label `target:beta` and the appropriate `target:YY00X-beta` (where YY00X is the targeted beta version).
- [x] Feature delivered into beta
  - ([OpenLiberty/release-manager](https://github.com/orgs/OpenLiberty/teams/release-manager)) adds label `release:YY00X-beta` (where YY00X is the first beta version that included the functionality).

### **Beta Blog** (Complete 1.5 weeks before beta eGA)
- [x] Beta blog issue created and populated using the [Open Liberty BETA blog post](https://github.com/OpenLiberty/open-liberty/issues/new/choose) template.
  - Add a link to the beta blog issue in the [Documents](#documents) section.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **GA**

A feature is ready to GA after it is Feature Complete and has obtained all necessary Focal Point Approvals. 

### **Feature Complete**
- [ ] Feature implementation and tests completed.
  - [ ] All PRs are merged.
  - [ ] All epic and child issues are closed.
  - [ ] All stop ship issues are completed.
- [ ] Legal: all necessary approvals granted.
- [ ] Translation: All messages translated or sent for translation for upcoming release 
- [ ] GA development complete and feature ready for inclusion in a GA release
  - Add label `target:ga` and the appropriate `target:YY00X` (where YY00X is the targeted GA version).
  - Inclusion in a release requires the completion of all Focal Point Approvals.

### **Focal Point Approvals** (Complete by Feature Complete Date)

These occur only after GA of this feature is requested (by adding a `target:ga` label).  GA of this feature may not occur until all approvals are obtained.

### **All Features**
- [ ] **FAT** All Tests complete and running successfully in SOE or N/A. ([OpenLiberty/fat-approvers](https://github.com/orgs/OpenLiberty/teams/fat-approvers))
  - Approver adds label `focalApproved:fat`.
- [x] **Demo** Demo is scheduled for an upcoming EOI or N/A. ([OpenLiberty/demo-approvers](https://github.com/orgs/OpenLiberty/teams/demo-approvers))
  - Approver adds label `focalApproved:demo`.
- [x] **Globalization** Translation and TVT are complete or N/A. ([OpenLiberty/globalization-approvers](https://github.com/orgs/OpenLiberty/teams/globalization-approvers))
  - Approver adds label `focalApproved:globalization`.

### **Design Approved Features**
- [x] **Accessibility** Accessibility testing completed or N/A. ([OpenLiberty/accessibility-approvers](https://github.com/orgs/OpenLiberty/teams/accessibility-approvers))
  - Approver adds label `focalApproved:accessibility`.
- [x] **APIs/Externals** Externals have been reviewed or N/A.  ([OpenLiberty/externals-approvers](https://github.com/orgs/OpenLiberty/teams/externals-approvers))
  - Approver adds label `focalApproved:externals` 
- [x] **ID** Documentation is complete or N/A. ([OpenLiberty/id-approvers](https://github.com/orgs/OpenLiberty/teams/id-approvers))
  - Approver adds label `focalApproved:id`.
  - > **_NOTE:_**  If only trivial documentation changes are required, you may reach out to the ID Feature Focal to request a `ID Required - Trivial` label.  Unlike features with regular ID requirement, those with `ID Required - Trivial` label do not have a hard requirement for a Design/UFO.
- [x] **Performance** Performance testing is complete or N/A. ([OpenLiberty/performance-approvers](https://github.com/orgs/OpenLiberty/teams/performance-approvers))
  - Approver adds label `focalApproved:performance`.
- [x] **Serviceability** Serviceability has been addressed or N/A. ([OpenLiberty/serviceability-approvers](https://github.com/orgs/OpenLiberty/teams/serviceability-approvers))
  - Approver adds label `focalApproved:sve`.
- [x] **STE** Skills Transfer Education chart deck is complete or N/A. ([OpenLiberty/ste-approvers](https://github.com/orgs/OpenLiberty/teams/ste-approvers))
  - Approver adds label `focalApproved:ste`.
- [x] **SVT** System Verification Test is complete or N/A. ([OpenLiberty/svt-approvers](https://github.com/orgs/OpenLiberty/teams/svt-approvers))
  - Approver adds label `focalApproved:svt`.

### **Remove Beta Fencing** (Complete by Feature Complete Date)
- [ ] Beta guards are removed, or N/A
  - Only after all necessary Focal Point Approvals have been granted.

### **GA Blog** (Complete by Feature Complete Date)
- [ ] GA Blog issue created and populated using the [Open Liberty GA release blog post](https://github.com/OpenLiberty/open-liberty/issues/new/choose) template.
  - Add a link to the GA Blog issue in the [Documents](#documents) section.

### **Post GA**
- [ ] Replace `target:YY00X` label with the appropriate `release:YY00X`. ([OpenLiberty/release-manager](https://github.com/orgs/OpenLiberty/teams/release-manager))

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Other Deliverables**

- [ ] **OL Guides** OL Guides assessment is complete or N/A. ([Yee-Kang Chang](https://github.com/yeekangc))
- [ ] **Standalone Feature Blog Post** A blog post specifically about your feature or N/A.  ([OpenLiberty/release-architect](https://github.com/orgs/OpenLiberty/teams/release-architect))
  - This should be strongly considered for larger or more prominent features.
  - Follow [instructions](https://github.com/OpenLiberty/blogs/tree/draft#writing-and-publishing-blog-posts-on-the-openlibertyio-blog) in the blogs repo.
- [ ] **WDT** Liberty Developer Tools work is complete or N/A. (Leonard Theivendra)

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

**Comment by Channyboy** [comment]
@OpenLiberty/globalization-approvers   N/A No Need for globalization. No translatable files changed.
@OpenLiberty/accessibility-approvers  N/A No accessibility concerns.
@OpenLiberty/externals-approvers  N/A: No external API changes or configuration changes.
@OpenLiberty/performance-approvers  N/A: Change does not incur additional perf (adds existing log record object to liberty output)
@OpenLiberty/svt-approvers  N/A: No additional SVT required?
@OpenLiberty/ste-approvers Does this need a STE doc?? (this only adds a new JSON field if a `throwable` is included as a param as part of JUL calls in an application.

**Comment by donbourne** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  UFO -- does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.

5.  Does this feature add any new metrics or emit any new JSON events?  If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?

**Comment by tngiang73** [comment]
@Channyboy: L2 confirmed STE is not needed. Thanks.

**Comment by Channyboy** [comment]
@donbourne 
1. N/A: no customer issues anticipated
2.a) The feature in its entirety was demonstrated. No problem paths are expected from the feature.
b) SVT team (Tanya and Steven)
c) yes
3) a/b) No SVT is needed as discussed with Brian from SVT.
4) WAS L3: Logging
5) yes. Changes to JSON ref doc at : https://github.com/OpenLiberty/docs/issues/5631

**Comment by donbourne** [comment]
@Channyboy , for 5 you mean that there are updates to the json log representation, and that you're updating the docs to cover that?

**Comment by Channyboy** [comment]
@donbourne  Yes. There will be changes to JSON field reference page. Issue at https://github.com/OpenLiberty/docs/issues/5631

**Comment by donbourne** [comment]
thanks, @Channyboy .

please also update the link in this issue to the UFO -- it appears the shared link has expired.  You may need to set the expiry for the shared link explicitly so it doesn't expire so quickly.

**Comment by chirp1** [comment]
David has created a draft of the documentation for issue https://github.com/OpenLiberty/docs/issues/5631 . Approving.
